common/ledger/blkstorage/blocks_itr.go

```diff
@@ -68,12 +69,32 @@ func (itr *blocksItr) Next() (ledger.QueryResult, error) {
 	if itr.closeMarker {
 		return nil, nil
 	}
+
+	cachedBlock, existsInCache := itr.mgr.cache.retrieveCachedBlock(itr.blockNumToRetrieve)
+	if existsInCache {
+		logger.Debugf("Retrieved block %d from ledger in-memory cache", itr.blockNumToRetrieve)
+		itr.cachedPrev = true
+		itr.blockNumToRetrieve++
+		return cachedBlock, nil
+	}
+
+	if itr.cachedPrev {
+		itr.cachedPrev = false
+		if itr.stream != nil {
+			itr.stream.close()
+			itr.stream = nil
+		}
+	}
```

Comment:
The only way to open the stream and then close it, is if you retrieved a block from the cache and then you got a cache miss. 

If you retrieve a block from the cache and then get a cache miss, it means you were pulling the "tail" of the cache (since it is continuous) and then the ledger is appended with so many transactions per second that the client that pulls the transactions can't keep up with the speed of blocks being appended to the ledger.

If the client is the peer or something equivalent to the peer that processed transactions, and it can't keep up with the ordering service, then we have a much bigger problem here as the effective latency of transactions will explode.

Reply:
Just for the sake of completeness - This could happen in the normal scenario as well as you first append the block to the file and then add it to the cache. Occasionally, the iterator may keep oscillating between the file and the cache. But, yes, that's not expected to be frequent at the regular transaction rates.

However, one note on code readability here. Why do you have this `if itr.cachedPrev {` block? Can't you simply close the stream inside the `if existsInCache {` block and get rid of `cachedPrev` concept altogether?

Reply:
Done